[Step3] 다대다 연관 관계 리팩터링

charllossdev opened the pull request:
안녕하세요 우빈님 👍

3단계 다대다 연관 관게 리뷰 요청드립니다.

수업강의를 듣고 1단계부터 다대다 양방향 조인관계로 개발하여 특별한 수정사항이 없었습니다.

혹시 놓친 부분이나 부족한 부분이 있다면 강력하게 피드백 부탁드릴게요.

많은 피드백 부탁드리겠습니다.
감사합니다 :)

charllossdev committed:
fix: 피드백 반영 @Getter 추가

charllossdev committed:
delete: subway 도메인 및 테스트 삭제

charllossdev committed:
refactor: position - line 양방관계 제거 리펙토링

charllossdev committed:
test: position 생성 테스트

charllossdev committed:
fix: member - favorite 양방향 관계 설정

charllossdev committed:
refactor: position 양방향 편의 메서드 추가

wbluke commented on src/main/java/jpa/favorite/Favorite.java:
```diff
@@ -1,30 +1,37 @@
 @Entity
 public class Favorite extends BaseTime {
 
+	@Getter
```
Entity 클래스에 @Getter를 달아도 될 것 같아요~

wbluke commented on src/main/java/jpa/line/Line.java:
```diff
@@ -40,10 +40,6 @@
 	@ManyToMany(mappedBy = "lines")
 	private final List<Station> stations = new ArrayList<>();
```
다대다 매핑에는 단점이 존재하기 때문에 사용하지 않고 다른 방법으로 풀어내어 설계합니다 (Step3 의 핵심이에요 ㅎㅎ)
어떤 단점이 존재할까요? :) [참고](https://ict-nroo.tistory.com/127)

(이미 Position 이라는 엔티티가 그 역할을 하고 있는것 같은데요. Position도 변경 필요한 부분이 있으니 한번 생각해보시길 바라요 🙂  )

wbluke commented on src/main/java/jpa/member/Member.java:
```diff
@@ -18,6 +17,7 @@
 @Entity
 public class Member extends BaseTime {
```
특별한 옵션값을 줄게 아니라면 @Column은 생략하셔도 됩니다 :)

wbluke commented on src/main/java/jpa/position/Position.java:
```diff
@@ -18,36 +18,35 @@
-	@ManyToOne
-	@JoinColumn
-	private Line line;
+	@ManyToOne(fetch = FetchType.LAZY)
```
fetchType 적용 잘 해주셨어요 ㅎㅎ
참고로 항상 적용하는게 만능은 아니고, 일단 LAZY로 설정했다가 도메인 로직을 추가하면서 즉시로딩이 필요한 곳이 어디일지 고민해주시면 좋을 것 같아요 :)

wbluke reviewed (changes requested):
피드백 반영 감사합니다 👍 
Step3 핵심 내용에 대해 고민과 리팩토링이 필요한 듯 해요 ㅎㅎ
코멘트 확인하시고 수정 후에 재요청 부탁드려요! 🙂

charllossdev committed:
style: 코드 정리

charllossdev committed:
refactor: 라인 생성 및 거리 추가 테스트 리펙토링

charllossdev committed:
refactor: 라인 생성 및 거리 추가 테스트 리펙토링

charllossdev committed:
refactor: ManyToMany 엔티티 연관관계 리펙토링

- ManyToMany 관계 리펙토링 이유
    + M:N 인 경우, JPA가 자동으로 M:N 관리 엔티티를 생성
- M : N 을 관리해주는 새로운 엔티티 생성

charllossdev committed:
fix: 전철 역 생성 테스트 수정

charllossdev committed:
fix: 전철 역 생성 테스트 수정

charllossdev committed:
fix: 위치 생성 테스트 수정

charllossdev committed:
fix: 유저, 즐겨찾기 테스트 수정

charllossdev committed:
refactor: 즐겨찾기 리펙토링

charllossdev committed:
feat: 거리 wrapping 객체 distance 생성

charllossdev committed:
test: 위치 엔티티 거리 wrapping 객체 테스트

charllossdev committed:
refactor: 라인 리펙토링

charllossdev commented:
안녕하세요 우빈님 :)
피드백을 받고 엄청나게 고민을 많이 했습니다.

그러던 중에 급한 마음에 인수테스트 과제를 먼저 시작했는데, 생각지도 못하게 문득 정답을 느끼게 됬습니다.

다대다 맵핑에 관해서 피드백과 참고자료 정말 감사합니다.
정리가 깔끔하게 되었으며, 많은 공부가 되었습니다!!!

추가로 질문이 있습니다.
지금 `Position` 엔티티와 `Station`이 다대일 단방향 관계에서 양방향 관계로 바꾸고 싶은데,
어떻게 해야 할지 몰라서 질문을 드립니다.

> `Position` 엔티티가 `Station`과 2개의 조인 단방향 관계를 가지는 요소를 양방향으로

wbluke commented on src/main/java/jpa/position/Position.java:
```diff
@@ -11,43 +15,53 @@
-	@ManyToOne
-	@JoinColumn
-	private Station station;
+	@ManyToOne(fetch = FetchType.LAZY)
+	@JoinColumn(nullable = false)
+	private Station upStation;
```
질문에 대한 답변 여기에 남길게요 ㅎㅎ
양방향으로 설정하고 싶다면 Station 쪽에서 @OneToMany로 mappedBy 설정을 하게 되는데요.
말그대로 반대편(Position)에 있는 필드를 지정하는 속성입니다.
결론적으로 mappedBy로 각각 upStation과 downStation을 지정하면 됩니다!
(정말 되는지 궁금하시면 테스트 짜서 실험해보셔도 좋을 것 같아요 ㅎㅎ)

charllossdev replied:
아 이해했습니다.

저는 한개의 필드만 만들어서 조인하려고 했는데, 
생각해보니 각각 지정해서 간단하게 해결했습니다.

감사합니다!!! 👍

wbluke commented on src/main/java/jpa/position/Position.java:
```diff
@@ -11,43 +15,53 @@
+	@JoinColumn(nullable = false)
+	private Station downStation;
+
+	@Resource
+	private transient Distance distance;
```
Distance를 포장해주셨네요! 👍 
다만 @Resource, transient 키워드를 제거하고 @Embedded, @Embeddable 에 대해 학습한 뒤에 적용해 주세요 :)
[참고](https://cheese10yun.github.io/jpa-embedded/)

charllossdev replied:
헉 제가 잘못한거군요.... @Transient 를 사용하면 엔티티에 필드가 생기지 않네요...

@Embedded 로 수정하였습니다.

wbluke reviewed (changes requested):
다대다 관계에 대한 고민 잘 해주셨네요 👍 
질문에 대한 답변도 코멘트로 남겼습니다!
그리고 새롭게 적용해주신 부분에 대해서 짚고 넘어갈게 있어서, 확인 후에 다시 요청 주시면 merge 할게요 🙂

charllossdev committed:
fix: distance wrapper 엔티티 객체 생성

charllossdev committed:
fix: station - position 각 엔티티 조인

charllossdev committed:
test: postion <-> station 양방향 조인 테스트

wbluke commented on src/main/java/jpa/position/Distance.java:
```diff
@@ -1,11 +1,19 @@
 
-	@Getter
-	private final long distance;
+	@Column
```
특별한 속성이 없는 @Column은 붙이지 않아도 됩니다 :)

wbluke reviewed (approved):
피드백 반영 잘 해주셨네요 👍 
조금 길었던 JPA 미션 고생 많으셨어요!
남은 미션도 화이팅입니다 💪
